### PR TITLE
Add sorting strategy

### LINKED ISSUE #8: Support stable sorting
[Dart's sort method](https://api.flutter.dev/flutter/dart-core/List/sort.html) which is used by `package:sorted` is unstable.  Since this method is at the core of `package:sorted` it is not possible to use a stable sort together with `package:sorted`. I propose to add a possibility of using a stable sort.

## Proposal 1

Adding `{bool stable = false}` to `Sorted<T>.sorted` extension (with a default of `false` to not introduce any changes in behavior). Then, if stable is `true`, a stable sort would be used instead.

Standard library does not provide any stable sort algorithms. Here I see two possibilities: adding `package:collection` to dependencies which provides such algorithms but this would make `package:sorted` no longer dependency-less, or copying the implementation locally to `package:sorted` (it is not long nor complex). As for the specific implementation, we could use a merge sort which has an easy implementation but introduces $\mathcal O(n)$ space complexity (which is generally the case with stable sorts). Or if we want to get very fancy we could use some state-of-the-art stable sort implementation, like an adaptive sort ([example from rust std](https://doc.rust-lang.org/std/primitive.slice.html#impl-3)) which optimizes edge cases.

## Proposal 2

If proposal 1 is not satisfactory, instead `SortedRule` could just accept a custom sorting function, then the responsibility of space/runtime considerations and stableness/unstableness is left to the user.

Personally, I would prefer proposal 1 for its simplicity.

### PR DESCRIPTION
**Pull request by shilangyu**
Qs:
- `SortingStrategy` vs `SortStrategy`?
- Merge sort copied from flutter with a license notice: leave as is, just reimplement merge sort, or pull in collection as dependency?
- `SortingStrategy` has an API which forces the sort to be done in place. We could instead move the responsibility of cloning the list to `SortingStrategy` which would give more flexibility but might introduce unexpected in-place operations if not used carefully.

closes #8

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# 2.1.0
+
+- Add `SortingStrategy` which allows for customizing the sorting algorithm.
```
Comment (Albert221) [inline CHANGELOG.md:3]: Mention available strategies

File: lib/src/extension.dart [hunk 2]
```diff
@@ -1,21 +1,37 @@
...
-    if (rules.isEmpty) return List.of(this);
+  ///
+  /// A sorting strategy can be provided through [sortingStrategy] which
+  /// will perform the final sorting procedure. Defaults to [DefaultSortingStrategy].
+  /// Consider using [MergeSortingStrategy] if a stable sort is needed.
+  List<T> sorted(
+    List<SortedRule> rules, {
+    SortingStrategy<T>? sortingStrategy,
+  }) {
+    sortingStrategy ??= DefaultSortingStrategy<T>();
```
Comment (Albert221) [inline lib/src/extension.dart:18]: Move it to the parameters list as a default value. This class is const

File: lib/src/extension.dart [hunk 3]
```diff
@@ -1,21 +1,37 @@
+import 'package:sorted/src/sorting_strategies/default.dart';
+import 'package:sorted/src/sorting_strategies/merge.dart';
+import 'package:sorted/src/sorting_strategy.dart';
+
 import 'rule.dart';
 
 extension Sorted<T> on Iterable<T> {
-  /// Returns a sorted list according to the `rules`. Order matters, so if
+  /// Returns a sorted list according to the [rules]. Order matters, so if
```
Comment (Albert221) [inline lib/src/extension.dart:8]:
```suggestion
  /// Returns a sorted list according to the `rules`. Order matters, so if
```

Square brackets don't make valid links to method/function parameters

File: lib/src/extension.dart [hunk 4]
```diff
@@ -1,21 +1,37 @@
...
   /// first rule considers two items equal, next one will decide and so on.
-  List<T> sorted(List<SortedRule> rules) {
-    if (rules.isEmpty) return List.of(this);
+  ///
+  /// A sorting strategy can be provided through [sortingStrategy] which
```
Comment (Albert221) [inline lib/src/extension.dart:11]:
```suggestion
  /// A sorting strategy can be provided through `sortingStrategy` which
```

File: lib/src/sorting_strategies/default.dart [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import 'package:sorted/src/sorting_strategy.dart';
+
+/// A [SortingStrategy] which uses dart's default (unstable) sorting method
```
Comment (Albert221) [inline lib/src/sorting_strategies/default.dart:3]:
```suggestion
/// A [SortingStrategy] which uses Dart's default (unstable) sorting method.
```

File: lib/src/extension.dart [hunk 6]
```diff
@@ -1,21 +1,37 @@
+import 'package:sorted/src/sorting_strategies/default.dart';
+import 'package:sorted/src/sorting_strategies/merge.dart';
+import 'package:sorted/src/sorting_strategy.dart';
```
Comment (Albert221) [inline lib/src/extension.dart:3]: Let's use only relative imports inside `src/`

File: lib/src/sorting_strategies/merge.dart [hunk 7]
```diff
@@ -0,0 +1,300 @@
+import 'package:sorted/src/sorting_strategy.dart';
+
+/// A [SortingStrategy] which uses merge sort, a stable sorting algorithm
```
Comment (Albert221) [inline lib/src/sorting_strategies/merge.dart:3]:
```suggestion
/// A [SortingStrategy] which uses merge sort, a stable sorting algorithm.
```

File: lib/src/sorting_strategy.dart [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import 'sorting_strategies/default.dart';
+import 'sorting_strategies/merge.dart';
+
+/// Strategy performing the final sorting procedure given a list and a single comparator.
+///
+/// See also:
+///
+///  * [DefaultSortingStrategy], which uses dart's default (unstable) sorting method
```
Comment (Albert221) [inline lib/src/sorting_strategy.dart:8]:
```suggestion
///  * [DefaultSortingStrategy], which uses Dart's default (unstable) sorting method.
```

File: lib/src/sorting_strategy.dart [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+///  * [MergeSortingStrategy], which uses merge sort, a stable sorting algorithm
```
Comment (Albert221) [inline lib/src/sorting_strategy.dart:9]:
```suggestion
///  * [MergeSortingStrategy], which uses merge sort, a stable sorting algorithm.
```

File: test/sorting_strategy_test.dart [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import 'package:sorted/sorted.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('SortingStrategy', () {
+    group('DefaultSortingStrategy', () {
```
Comment (Albert221) [inline test/sorting_strategy_test.dart:6]: Please split those tests in two files, for each sorting strategy

**Review by Albert221 — CHANGES_REQUESTED** [review]
* Let's use `mergeSort` from the `collection` package. Thanks to this, Flutter will add its LICENSE to its license registry and we don't have to deal with having 3rd party code in the repository.
* `SortingStrategy` name is OK for me.
* Sorting method in Dart mutates the passed object. This looks like a convention, where `sort` would mutate and `sorted` would be pure. I'm OK with how it mutates the list now

File: lib/src/extension.dart (hunk 2 above)
Comment (shilangyu) [reply lib/src/extension.dart:18]: `T` is not const so it can't be constructed as const

File: lib/src/sorting_strategy.dart [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+import 'sorting_strategies/merge.dart';
+
+/// Strategy performing the final sorting procedure given a list and a single comparator.
+///
+/// See also:
+///
+///  * [DefaultSortingStrategy], which uses Dart's default (unstable) sorting method.
+///  * [MergeSortingStrategy], which uses merge sort, a stable sorting algorithm.
+abstract class SortingStrategy<T> {
+  /// Sorts [list] in place given a [comparator]
```
Comment (Albert221) [inline lib/src/sorting_strategy.dart:11]:
```suggestion
  /// Sorts [list] in place given a [comparator].
```

**Review by Albert221 — APPROVED** [review]

**Pull request merged**